src/use-swr.ts

```diff
@@ -57,34 +58,40 @@ const mutate: mutateInterface = function(key, data, shouldRevalidate = true) {
   }
 }
 
-function useSWR<Data = any, Error = any>(
-  key: keyInterface
-): responseInterface<Data, Error>
-function useSWR<Data = any, Error = any>(
+function useSWR<
+  Data = any,
+  Error = any,
+  Fn extends fetcherFn<Data> = fetcherFn<Data>
+>(
   key: keyInterface,
+  fn?: Fn,
   config?: ConfigInterface<Data, Error>
 ): responseInterface<Data, Error>
-function useSWR<Data = any, Error = any>(
+function useSWR<
+  Data = any,
+  Error = any,
+  Fn extends fetcherFn<Data> = fetcherFn<Data>
+>(
   key: keyInterface,
-  fn?: Function,
+  fn?: Fn,
+  deps?: Parameters<Fn>,
```

Comment:
The react team recommends against having a custom `deps` as the lint-rule will not pick these up automatically. https://twitter.com/dan_abramov/status/1140316424898043904 The recommended way is to pass in a function wrapped via `useMemo` or `useCallback`.

Reply:
Also `deps` being not the last argument would be a very surprising API-design since the deps array is usually the last array in hooks.

Reply:
This isnt a "custom" `deps`. Its just an array. `Parameters` is a typescript type to put the arguments of a function in a tuple.

the eslint plugin for exhaustive deps is expecitng `deps` to be immediately after the function. Hands may be tied here for that